### DISCUSSION TITLE
AST-node-like exports and building queries with them

### DISCUSSION THREAD

**Discussion by edezhic** [tursodatabase/turso#580] [Ideas]
Hey, I'm building an app framework([prest](https://prest.blog/)) and I want an embedded rust-based database integrated into it. As of now I'm rolling with [gluesql](https://github.com/gluesql/gluesql), but after the limbo announcement I started thinking that it might make sense to switch. One of the things I like about gluesql is [ast builder](https://gluesql.org/docs/0.16.0/ast-builder/intro), especially rusty syntax + type checking and injection-free nature of it. As a bonus it also removes the need to parse queries. Do you plan anything like that? To export ast nodes in your public api and allow building/executing queries from them.

**Comment by penberg** [comment]
@edezhic Limbo follows the same architecture as SQLite of converting the SQL queries into bytecode programs very early on. The parser does generate some AST but probably want to flatten it to improve prepare() performance. So It's not clear Limbo will have the kind of AST builder you're looking for. But maybe @jussisaurio is a better person to give an answer here.

**Reply by Wukuyon** [reply]
A statement-builder API that uses functions or methods to create statements, instead of parsing SQL strings, would allow applications to bypass the SQL parser and more-securely create dynamic queries from structured user data, without needing to worry about invalid SQL syntax due to string concatenation or interpolation. It could also leverage any compile-time type checker in the host language.

Notably, [D. Richard Hipp said that SQLite ASTs are unlikely to ever be publicly stabilized—and that non-SELECT statements don't even have get full ASTs](https://sqlite.org/forum/forumpost/b5d21f5d91cc09455b7150af63a69e860d977d910c82e7ed77e528182e6201fc), so I think a rewrite has a big opportunity here to create a query-builder API where SQLite will not.

(This was originally from #911, but I moved it here to make that thread focus on compile-time query typing.)

**Comment by Wukuyon** [comment]
A shiny example of a type-safe SQL query builder is [Kysely for TypeScript](https://kysely.dev). With Kysely, you can construct queries with static typing and TypeScript-driven editor autocompletion. It would be fantastic if Turso built a similar feature more deeply into its engine.
